RPC: Accept options as named-only parameters

ryanofsky opened the pull request:
Allow RPC methods which take an `options` parameter (`importmulti`, `listunspent`, `fundrawtransaction`, `bumpfee`, `send`, `sendall`, `walletcreatefundedpsbt`, `simulaterawtransaction`), to accept the options as named parameters, without the need for nested JSON objects.

This makes it possible to make calls like:

```sh
src/bitcoin-cli -named bumpfee txid fee_rate=10
```

instead of

```sh
src/bitcoin-cli -named bumpfee txid options='{"fee_rate": 10}'
```

RPC help is also updated to show options as top level named arguments instead of as nested objects.


<details><summary>diff</summary>
<p>

```diff
@@ -15,16 +15,17 @@
 
 Arguments:
 1. txid                           (string, required) The txid to be bumped
-2. options                        (json object, optional)
+2. options                        (json object, optional) Options object that can be used to pass named arguments, listed below.
+                 
+Named Arguments:
-     {
-       "conf_target": n,          (numeric, optional, default=wallet -txconfirmtarget) Confirmation target in blocks
+conf_target                       (numeric, optional, default=wallet -txconfirmtarget) Confirmation target in blocks
 
-       "fee_rate": amount,        (numeric or string, optional, default=not set, fall back to wallet fee estimation) 
+fee_rate                          (numeric or string, optional, default=not set, fall back to wallet fee estimation) 
                                   Specify a fee rate in sat/vB instead of relying on the built-in fee estimator.
                                   Must be at least 1.000 sat/vB higher than the current transaction fee rate.
                                   WARNING: before version 0.21, fee_rate was in BTC/kvB. As of 0.21, fee_rate is in sat/vB.
                                   
-       "replaceable": bool,       (boolean, optional, default=true) Whether the new transaction should still be
+replaceable                       (boolean, optional, default=true) Whether the new transaction should still be
                                   marked bip-125 replaceable. If true, the sequence numbers in the transaction will
                                   be left unchanged from the original. If false, any input sequence numbers in the
                                   original transaction that were less than 0xfffffffe will be increased to 0xfffffffe
@@ -32,11 +33,10 @@
                                   still be replaceable in practice, for example if it has unconfirmed ancestors which
                                   are replaceable).
                                   
-       "estimate_mode": "str",    (string, optional, default="unset") The fee estimate mode, must be one of (case insensitive):
+estimate_mode                     (string, optional, default="unset") The fee estimate mode, must be one of (case insensitive):
                                   "unset"
                                   "economical"
                                   "conservative"
-     }
 
 Result:
 {                    (json object)
```

</p>
</details>


**Review suggestion:** To understand this PR, it is probably easiest to review the commits in reverse order because the last commit shows the external API changes, the middle commit shows the internal API changes, and the first commit contains the low-level implementation.

ryanofsky committed:
RPC: Add add OBJ_NAMED_PARAMS type

OBJ_NAMED_PARAMS type works the same as OBJ type except it registers the object
keys to be accepted as top-level named-only RPC parameters. Generated
documentation also lists the object keys seperately in a new "Named arguments"
section of help text.

Named-only RPC parameters have the same semantics as python keyword-only
arguments (https://peps.python.org/pep-3102/). They are always required to be
passed by name, so they don't affect interpretation of positional arguments,
and aren't affected when positional arguments are added or removed.

The new OBJ_NAMED_PARAMS type is used in the next commit to make it easier to
pass options values to various RPC methods.

Co-authored-by: Andrew Chow <github@achow101.com>

ryanofsky committed:
test: Update python tests to use named parameters instead of options objects

ryanofsky committed:
RPC: Allow RPC methods accepting options to take named parameters

Co-authored-by: Andrew Chow <github@achow101.com>

ryanofsky commented:
This is a draft because it needs more unit tests and is not tested very much yet in general. Also probably should have release notes and updates to generated documentation.

luke-jr commented:
Concept ACK. Another approach might be #17356. Not sure which is better (17356 is probably more readable code?)

ryanofsky commented:
@MarcoFalke you may have thoughts on `RPCHelpMan` changes in this PR. I didn't make any changes to generated documentation yet, but I'm thinking adding a new "Named arguments" section between the existing "Arguments" and "Result" sections:

<pre>
Arguments:
1. arg1 (type) description
2. arg2 (type) description

<b>Named arguments:
option1 (type) description
option2 (type) description</b>

Result:
result (type) description
</pre>

For RPC methods that take explicit `options` arguments (existing RPC methods like like `importmulti`, `listunspent`, etc), the documentation lines for those arguments could look like:

```
3. options (json object, optional) Options object that can be used to pass named arguments
```

luke-jr commented:
Concept ACK on a simple RPC help for "options", but I wonder if there's a better way to avoid duplicating all positional argument docs... Unless your thought was to have only non-positional ones under the `Named arguments` list?

ryanofsky commented:
> Concept ACK on a simple RPC help for "options", but I wonder if there's a better way to avoid duplicating all positional argument docs... Unless your thought was to have only non-positional ones under the `Named arguments` list?

Right, no duplication. Arguments that can be named and positional shown with names and numbers in "Arguments" section. Arguments that can only be passed by name shown without numbers in the "Named arguments"

ryanofsky commented:
Rebased e1b5be5d5a3d3a0ea5a53fa1acd47128efa75bb0 -> fae439b117bb90b6288b2e9b807e8219439cd726 ([`pr/nonly.3`](https://github.com/ryanofsky/bitcoin/commits/pr/nonly.3) -> [`pr/nonly.4`](https://github.com/ryanofsky/bitcoin/commits/pr/nonly.4), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/nonly.3-rebase..pr/nonly.4)) after base PR #19762 merge. Also added unit tests and implemented changes to generated documentation discussed above.
Rebased fae439b117bb90b6288b2e9b807e8219439cd726 -> cea5e37d4ec34d62c413c27775b7bafeaebff6c4 ([`pr/nonly.4`](https://github.com/ryanofsky/bitcoin/commits/pr/nonly.4) -> [`pr/nonly.5`](https://github.com/ryanofsky/bitcoin/commits/pr/nonly.5), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/nonly.4-rebase..pr/nonly.5)) due to conflict with #26601.

903803 commented:
src/rpc/server.cpp

aureleoules commented:
Concept ACK

ryanofsky commented:
Updated cea5e37d4ec34d62c413c27775b7bafeaebff6c4 -> fb39d5a14a8a4ed22cf6cad523a51cf709f95650 ([`pr/nonly.5`](https://github.com/ryanofsky/bitcoin/commits/pr/nonly.5) -> [`pr/nonly.6`](https://github.com/ryanofsky/bitcoin/commits/pr/nonly.6), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/nonly.5..pr/nonly.6)) adding release notes, internal documentation, and making a few simplifications
Rebased fb39d5a14a8a4ed22cf6cad523a51cf709f95650 -> ec11f06e472eb7e1f5654ab8537feff0f17065a5 ([`pr/nonly.6`](https://github.com/ryanofsky/bitcoin/commits/pr/nonly.6) -> [`pr/nonly.7`](https://github.com/ryanofsky/bitcoin/commits/pr/nonly.7), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/nonly.6-rebase..pr/nonly.7)) due to conflict with #26628

ryanofsky commented:
Rebased d936e991a63db798c9ac68238c5c48d42cdd65c7 -> 7fd1401d8116323adfa2a87bbcf6ea41437cd0fa ([`pr/nonly.10`](https://github.com/ryanofsky/bitcoin/commits/pr/nonly.10) -> [`pr/nonly.11`](https://github.com/ryanofsky/bitcoin/commits/pr/nonly.11), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/nonly.10-rebase..pr/nonly.11)) due to conflict with #25344

TheCharlatan commented:
Concept ACK

This has been annoying me for years. I think there is one more `options` field that could be converted to `OBJ_NAMED_PARAMS` in the `scanblocks` call. 

There is also the `template_request` argument in `getblocktemplate` and the `solving_data` argument in `fundrawtransaction` that could be converted as well, but they contain a bunch of array arguments, so I'm not sold if this would actually increase ergonomics.

The tests now all seem to be using the named arguments. Are there tests ensuring the current option arguments can still be read?

ryanofsky commented:
Rebased 7fd1401d8116323adfa2a87bbcf6ea41437cd0fa -> 927a910cbe6c96d60c6a8b05baad179ef3f61f1c ([`pr/nonly.11`](https://github.com/ryanofsky/bitcoin/commits/pr/nonly.11) -> [`pr/nonly.12`](https://github.com/ryanofsky/bitcoin/commits/pr/nonly.12), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/nonly.11-rebase..pr/nonly.12)) due to conflict with #26642.

ajtowns commented:
There are a two cases where the same name is valid both as a positional argument and in the `options` object: `send` and `sendall` both have `conf_target`, `estimate_mode` and `fee_rate` as named/positional parameters as well as `options` members (`fee_rate` explicitly, the others via `FundTxDoc`). Both those RPCs are marked as "EXPERIMENTAL", so perhaps it would be good to change their API now?

I believe:

```diff
--- a/src/rpc/util.cpp
+++ b/src/rpc/util.cpp
@@ -493,6 +493,14 @@ RPCHelpMan::RPCHelpMan(std::string name, std::string description, std::vector<RP
         for (const std::string& name : names) {
             CHECK_NONFATAL(named_args.insert(name).second);
         }
+        if (arg.m_type == RPCArg::Type::OBJ_NAMED_PARAMS) {
+            for (const auto& inner : arg.m_inner) {
+                std::vector<std::string> inner_names = SplitString(inner.m_names, '|');
+                for (const std::string& inner_name : inner_names) {
+                    CHECK_NONFATAL(named_args.insert(inner_name).second);
+                }
+            }
+        }
         // Default value type should match argument type only when defined
         if (arg.m_fallback.index() == 2) {
             const RPCArg::Type type = arg.m_type;
```

would then let you catch these sorts duplicates as soon as  you run `bitcoind` after introducing them.

ajtowns commented:
Approach ACK

(just for the record)

ryanofsky commented:
re: https://github.com/bitcoin/bitcoin/pull/26485#issuecomment-1507916357

@ajtowns I like the safety check you suggested, and agree it could catch some programming errors. But I also think what `send` and `sendall` functions are trying to do to provide compatibility is reasonable, and adding the check would limit what they are able to do.

`send` and `sendall` currently accept `conf_target`, `estimate_mode` and `fee_rate` values as positional parameters (1), or as named parameters (2), or as options (3), and forbidding duplicate option/parameter names would require dropping support for either for (1) or (3). Dropping support for (1) would make these methods not backwards compatible and less convenient to call and dropping support for (3) would make them less consistent with `fundrawtransaction` and `walletcreatefundedpsbt` which take these values as options.

I do think it could make sense to add a check for duplicate options/named parameter fields in the future, but I don't see a way of doing it now without breaking things, and the having the extra safety check for programming errors doesn't seem critical here.

But if you do have a more specific suggestion for how to change the send/sendall functions, I'd definitely be curious. Maybe there is another way out of this tradeoff that I'm not thinking of.

ajtowns commented:
They're both experimental RPCs, so the fact that dropping support for them as a positional param (1) is a backwards incompatibility doesn't seem terribly important? That we can do positional+named params easily now also seems like it doesn't make things very inconvenient?

Could add an `RPCArgOptions{.also_positional=true}` to mark the `options` members that are intentionally duplicating positional arguments.

ryanofsky committed:
rpc: Add check for unintended option/parameter name clashes

Also add flag to allow RPC methods that intendionally accept options and
parameters with the same name bypass the check.

Check and flag were suggested by ajtowns
https://github.com/bitcoin/bitcoin/pull/26485#issuecomment-1507916357

Co-authored-by: Anthony Towns <aj@erisian.com.au>

ajtowns commented on src/rpc/util.cpp:
```diff
@@ -493,6 +493,14 @@ RPCHelpMan::RPCHelpMan(std::string name, std::string description, std::vector<RP
+            for (const auto& inner : arg.m_inner) {
+                std::vector<std::string> inner_names = SplitString(inner.m_names, '|');
+                for (const std::string& inner_name : inner_names) {
+                    CHECK_NONFATAL(named_args.insert(inner_name).second || inner.m_opts.also_positional);
```
I think this should be:

```c++
if (!inner.m_opts.also_positional) {
    CHECK_NONFATAL(named_args.insert(inner_name).second);
}
```

Otherwise a duplicate positional parameter after the `OBJ_NAMED_PARAMS` entry will give an error.

ryanofsky reviewed (commented):
Updated 284173937d156e1c0f291abe0b82274468d242e8 -> eaee226a17546b93245ca1435e4b468c368d9e86 ([`pr/nonly.13`](https://github.com/ryanofsky/bitcoin/commits/pr/nonly.13) -> [`pr/nonly.14`](https://github.com/ryanofsky/bitcoin/commits/pr/nonly.14), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/nonly.13..pr/nonly.14)) making check for duplicate parameter names less strict and adding more test coverage for the check

ajtowns commented:
ACK eaee226a17546b93245ca1435e4b468c368d9e86

Failing test looks like it was fixed in #27340 already.

ismaelsadeeq commented on src/rpc/server.cpp:
```diff
@@ -417,7 +417,9 @@ static inline JSONRPCRequest transformNamedArguments(const JSONRPCRequest& in, c
+    UniValue options{UniValue::VOBJ};
+    for (const auto& [argNamePattern, named_only]: argNames) {
         std::vector<std::string> vargNames = SplitString(argNamePattern, '|');
         auto fr = argsIn.end();
```
Concept ACK
Nit: As mentioned in review club, it would be nice to use a more descriptive variable name for the fr variable, considering your change is within the `transformNamedArguments` method. like find_result or a similar alternative.
```suggestion
        auto find_result = argsIn.end();
```

achow101 commented on src/rpc/server.cpp:
```diff
@@ -426,30 +428,58 @@ static inline JSONRPCRequest transformNamedArguments(const JSONRPCRequest& in, c
             if (out.params.empty()) initial_hole_size = hole;
         }
+
+        // If named input parameter "fr" is present, push it onto out.params. If
+        // options are present, push them onto out.params. If both are present,
+        // throw an error.
+        if (fr != argsIn.end()) {
+            if (!options.empty()) {
+                throw JSONRPCError(RPC_INVALID_PARAMETER, "Parameter " + fr->first + " conflicts with parameter " + options.getKeys().front());
+            }
```
In 411485082c22b86e1224f60534fccf1e2bb8e8f3 "RPC: Add add OBJ_NAMED_PARAMS type"

This seems somewhat fragile as it disallows having any arguments after the options object. While I would not expect that such arguments would be intentionally added to any RPC, it's possible that one could be added accidentally. Furthermore, the tests would not necessarily catch this issue if they do not use the named-only parameters.

achow101 commented:
ACK 2808c33bae95a0d2516a6e9a550032af142a04de

I find `translateNamedArguments` somewhat difficult to read, but it seems like there isn't really much that can be done to make it easier to comprehend without completely overhauling it. I've ended up doing that in #27788.

ajtowns commented:
reACK 2808c33bae95a0d2516a6e9a550032af142a04de

(I wouldn't expect bitcoin-cli to convert the types)

achow101 commented:
> (I wouldn't expect bitcoin-cli to convert the types)

Then this ends up being unusable, e.g.:

```
$ src/bitcoin-cli -regtest -rpcwallet=funds -named send '[{"bcrt1q4ul7y0p3p9d6sx6atnq9kucxng7jpde9tqfrt6":1}]' change_position=0
error code: -3
error message:
JSON value of type null is not of expected type string
```

ajtowns commented:
> > (I wouldn't expect bitcoin-cli to convert the types)
> 
> Then this ends up being fairly useless, e.g.:
> 
> ```
> $ src/bitcoin-cli -regtest -rpcwallet=funds -named send '[{"bcrt1q4ul7y0p3p9d6sx6atnq9kucxng7jpde9tqfrt6":1}]' change_position=0
> error code: -3
> error message:
> JSON value of type null is not of expected type string
> ```

Do you mean `JSON value of type string for field change_position is not of expected type number` ? (If not, I can't duplicate your error...)

Yeah, that does suck, and I don't see an obvious fix -- presumably we'd want named-only parameters to be treated as json rather than a string (unless overridden), but if it's not overridden, then bitcoin-cli has no way of knowing if `change_position` here is a positional param or a named one...

achow101 commented:
> Do you mean `JSON value of type string for field change_position is not of expected type number` ? (If not, I can't duplicate your error...)

Oops, yes, I don't think I was on the HEAD commit.

> Yeah, that does suck, and I don't see an obvious fix -- presumably we'd want named-only parameters to be treated as json rather than a string (unless overridden), but if it's not overridden, then bitcoin-cli has no way of knowing if `change_position` here is a positional param or a named one...

The obvious fix is to update the table `src/rpc/client.cpp` to contain all of the named-only parameters as we do for all the other parameters. It's just a bit tedious. They could just use the same position index as the existing `options` to avoid converting any other positionals.

But also we should probably figure out a better way to deal with the conversions in general.

ryanofsky commented:
> But also we should probably figure out a better way to deal with the conversions in general.

I think it was an unfortunate decision to hardcode parameter information into the client, and could think of various alternatives.

One alternative could be to avoid the need to process arguments on the client side at all and and add a new `exec` RPC method that takes an `args` array of strings and an optional `stdin` string parameter to figure out on the server side where there is full type information what RPC method to call and how to convert the string args to JSON. An `exec` method could also do more interpretation like #20273 if that is desirable. The bitcoin client could have a new `-exec` or `-dumb` option to invoke the `exec` RPC instead of trying to convert arguments to JSON itself. Or `-exec` could just be the default client behavior. 

Another alternative could be to have the client download parameter information from the server instead of hardcoding it, maybe through a schema specification like https://open-rpc.org/getting-started. Probably the client would need to cache this information to avoid a performance penalty, and it could be complicated deciding when to invalidate the cache. Though in the most common case when the client is using a cookie file, the server could just write an `rpc-schema.json` file next to the cookie and the client could just use that without downloading anything.

achow101 commented:
> One alternative could be to avoid the need to process arguments on the client side at all and and add a new `exec` RPC method that takes an `args` array of strings and an optional `stdin` string parameter to figure out on the server side where there is full type information what RPC method to call and how to convert the string args to JSON. An `exec` method could also do more interpretation like #20273 if that is desirable. The bitcoin client could have a new `-exec` or `-dumb` option to invoke the `exec` RPC instead of trying to convert arguments to JSON itself. Or `-exec` could just be the default client behavior.

Could we just have the server automatically convert the arguments as needed? This can still be done before the RPC execution since we have type information in the RPCHelpMan.

achow101 commented:
@ryanofsky Are you planning on making any further changes here?

ryanofsky commented:
> @ryanofsky Are you planning on making any further changes here?

I'm not working on anything. The only suggestion I saw was to rename the `fr` variable, but that is a preexisting variable so would prefer not to change that here.

achow101 commented:
Ah okay, I wasn't sure if you were going to add the named-only parameters to the conversion table, but based on the above discussion, it sounds like we're going to move towards eliminating the table in a followup.

ACK 2808c33bae95a0d2516a6e9a550032af142a04de

ryanofsky commented:
> I wasn't sure if you were going to add the named-only parameters to the conversion table

Sorry, I just missed the part of your earlier comment where you suggested to this. I didn't realize the conversion table could handle parameters not passed by position. Should be fixed now though.

Updated 2808c33bae95a0d2516a6e9a550032af142a04de -> 5559ad2c69ef82c18843b9214e5ba3974834a1ae ([`pr/nonly.15`](https://github.com/ryanofsky/bitcoin/commits/pr/nonly.15) -> [`pr/nonly.16`](https://github.com/ryanofsky/bitcoin/commits/pr/nonly.16), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/nonly.15..pr/nonly.16)) with rpc/client.cpp entries

achow101 commented:
It looks like the conversion table is missing the arguments that take amounts; these also need to be converted. The test is also failing as the named only arguments are not being output in the `dump_all_command_conversions`, the regex can't handle negatives, and `psbt` has mismatched conversions.

The following diff should fix those issues:

```diff
diff --git a/src/rpc/client.cpp b/src/rpc/client.cpp
index be4d777169..7bd66ebbd2 100644
--- a/src/rpc/client.cpp
+++ b/src/rpc/client.cpp
@@ -101,7 +101,10 @@ static const CRPCConvertParam vRPCConvertParams[] =
     { "listunspent", 2, "addresses" },
     { "listunspent", 3, "include_unsafe" },
     { "listunspent", 4, "query_options" },
+    { "listunspent", -1, "minimumAmount" },
+    { "listunspent", -1, "maximumAmount" },
     { "listunspent", -1, "maximumCount" },
+    { "listunspent", -1, "minimumSumAmount" },
     { "listunspent", -1, "include_immature_coinbase" },
     { "getblock", 1, "verbosity" },
     { "getblock", 1, "verbose" },
@@ -134,6 +137,8 @@ static const CRPCConvertParam vRPCConvertParams[] =
     { "fundrawtransaction", -1, "changePosition"},
     { "fundrawtransaction", -1, "includeWatching"},
     { "fundrawtransaction", -1, "lockUnspents"},
+    { "fundrawtransaction", -1, "fee_rate"},
+    { "fundrawtransaction", -1, "feeRate"},
     { "fundrawtransaction", -1, "subtractFeeFromOutputs"},
     { "fundrawtransaction", -1, "input_weights"},
     { "fundrawtransaction", -1, "conf_target"},
@@ -151,6 +156,8 @@ static const CRPCConvertParam vRPCConvertParams[] =
     { "walletcreatefundedpsbt", -1, "changePosition"},
     { "walletcreatefundedpsbt", -1, "includeWatching"},
     { "walletcreatefundedpsbt", -1, "lockUnspents"},
+    { "walletcreatefundedpsbt", -1, "fee_rate"},
+    { "walletcreatefundedpsbt", -1, "feeRate"},
     { "walletcreatefundedpsbt", -1, "subtractFeeFromOutputs"},
     { "walletcreatefundedpsbt", -1, "conf_target"},
     { "walletcreatefundedpsbt", -1, "replaceable"},
@@ -186,6 +193,7 @@ static const CRPCConvertParam vRPCConvertParams[] =
     { "send", -1, "maxconf"},
     { "send", -1, "add_to_wallet"},
     { "send", -1, "change_position"},
+    { "send", -1, "fee_rate"},
     { "send", -1, "include_watching"},
     { "send", -1, "inputs"},
     { "send", -1, "locktime"},
@@ -200,6 +208,7 @@ static const CRPCConvertParam vRPCConvertParams[] =
     { "sendall", 3, "fee_rate"},
     { "sendall", 4, "options" },
     { "sendall", -1, "add_to_wallet"},
+    { "sendall", -1, "fee_rate"},
     { "sendall", -1, "include_watching"},
     { "sendall", -1, "inputs"},
     { "sendall", -1, "locktime"},
@@ -244,10 +253,12 @@ static const CRPCConvertParam vRPCConvertParams[] =
     { "gettxspendingprevout", 0, "outputs" },
     { "bumpfee", 1, "options" },
     { "bumpfee", -1, "conf_target"},
+    { "bumpfee", -1, "fee_rate"},
     { "bumpfee", -1, "replaceable"},
     { "bumpfee", -1, "outputs"},
     { "psbtbumpfee", 1, "options" },
     { "psbtbumpfee", -1, "conf_target"},
+    { "psbtbumpfee", -1, "fee_rate"},
     { "psbtbumpfee", -1, "replaceable"},
     { "psbtbumpfee", -1, "outputs"},
     { "logging", 0, "include" },
diff --git a/src/rpc/util.cpp b/src/rpc/util.cpp
index c35706ce24..2a1061d49f 100644
--- a/src/rpc/util.cpp
+++ b/src/rpc/util.cpp
@@ -706,17 +706,30 @@ std::string RPCHelpMan::ToString() const
 UniValue RPCHelpMan::GetArgMap() const
 {
     UniValue arr{UniValue::VARR};
+
+    auto push_back_arg_info = [&arr](const std::string& rpc_name, int pos, const std::string& arg_name, const RPCArg::Type& type) {
+        UniValue map{UniValue::VARR};
+        map.push_back(rpc_name);
+        map.push_back(pos);
+        map.push_back(arg_name);
+        map.push_back(type == RPCArg::Type::STR ||
+                      type == RPCArg::Type::STR_HEX);
+        arr.push_back(map);
+    };
+
     for (int i{0}; i < int(m_args.size()); ++i) {
         const auto& arg = m_args.at(i);
         std::vector<std::string> arg_names = SplitString(arg.m_names, '|');
         for (const auto& arg_name : arg_names) {
-            UniValue map{UniValue::VARR};
-            map.push_back(m_name);
-            map.push_back(i);
-            map.push_back(arg_name);
-            map.push_back(arg.m_type == RPCArg::Type::STR ||
-                          arg.m_type == RPCArg::Type::STR_HEX);
-            arr.push_back(map);
+            push_back_arg_info(m_name, i, arg_name, arg.m_type);
+            if (arg.m_type == RPCArg::Type::OBJ_NAMED_PARAMS) {
+                for (const auto& inner : arg.m_inner) {
+                    std::vector<std::string> inner_names = SplitString(inner.m_names, '|');
+                    for (const std::string& inner_name : inner_names) {
+                        push_back_arg_info(m_name, -1, inner_name, inner.m_type);
+                    }
+                }
+            }
         }
     }
     return arr;
diff --git a/test/functional/rpc_help.py b/test/functional/rpc_help.py
index 7acc3cbbd5..dbd4b3047e 100755
--- a/test/functional/rpc_help.py
+++ b/test/functional/rpc_help.py
@@ -32,7 +32,7 @@ def process_mapping(fname):
                 if line.startswith('};'):
                     in_rpcs = False
                 elif '{' in line and '"' in line:
-                    m = re.search(r'{ *("[^"]*"), *([0-9]+) *, *("[^"]*") *},', line)
+                    m = re.search(r'{ *("[^"]*"), *(-?[0-9]+) *, *("[^"]*") *},', line)
                     assert m, 'No match to table expression: %s' % line
                     name = parse_string(m.group(1))
                     idx = int(m.group(2))
@@ -85,8 +85,8 @@ class HelpRpcTest(BitcoinTestFramework):
 
         for argname, convert in converts_by_argname.items():
             if all(convert) != any(convert):
-                # Only allow dummy to fail consistency check
-                assert argname == 'dummy', ('WARNING: conversion mismatch for argument named %s (%s)' % (argname, list(zip(all_methods_by_argname[argname], converts_by_argname[argname]))))
+                # Only allow dummy and psbt to fail consistency check
+                assert argname in ['dummy', "psbt"], ('WARNING: conversion mismatch for argument named %s (%s)' % (argname, list(zip(all_methods_by_argname[argname], converts_by_argname[argname]))))
 
     def test_categories(self):
         node = self.nodes[0]
```

achow101 commented:
> I'm surprised amounts need to be converted, though

It actually may just be the test that checks the conversion table that requires it since `dump_all_command_conversions` marks amounts as needed to be converted. However converting amounts is also consistent with all of the other RPCs that accept amounts as arguments, even prior to that test being added.

achow101 commented:
ACK 2cd28e9fef5dd743bcd70025196ee311fcfdcae4

Only changes since last are the suggested conversion table changes
